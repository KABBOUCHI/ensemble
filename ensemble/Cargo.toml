[package]
name = "ensemble"
version = "0.0.0"
readme = { workspace = true }
edition = { workspace = true }
authors = { workspace = true }
license = { workspace = true }
repository = { workspace = true }
description = "A Laravel-inspired ORM"

[dependencies]
chrono = "0.4.26"
thiserror = "1.0.44"
ensemble_derive = { version = "0", path = "../ensemble_derive" }
