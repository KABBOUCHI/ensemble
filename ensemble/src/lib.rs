#![warn(clippy::all, clippy::pedantic, clippy::nursery)]

#[doc(hidden)]
pub use async_trait::async_trait;
use builder::Builder;
use serde::{de::DeserializeOwned, Serialize};
use std::fmt::Display;

pub mod builder;
mod connection;
pub mod migrations;
pub mod query;
pub mod types;
mod value;
pub use connection::setup;
pub use ensemble_derive::Model;

#[async_trait]
pub trait Model: DeserializeOwned + Serialize + Sized + Send + Sync {
    /// The type of the primary key for the model.
    type PrimaryKey: Display + DeserializeOwned + Serialize + Send + Sync;

    /// The name of the table for the model
    const TABLE_NAME: &'static str;

    /// The name of the primary key field for the model.
    const PRIMARY_KEY: &'static str;

    /// Returns the names of the fields for the model.
    fn keys() -> Vec<&'static str>;

    /// Returns the value of the model's primary key.
    fn primary_key(&self) -> &Self::PrimaryKey;

    /// Get all of the models from the database.
    ///
    /// # Errors
    ///
    /// Returns an error if the query fails, or if a connection to the database cannot be established.
    async fn all() -> Result<Vec<Self>, query::Error> {
        query::all().await
    }

    /// Find a model by its primary key.
    ///
    /// # Errors
    ///
    /// Returns an error if the model cannot be found, or if a connection to the database cannot be established.
    async fn find(key: Self::PrimaryKey) -> Result<Self, query::Error>;

    /// Insert a new model into the database.
    ///
    /// # Errors
    ///
    /// Returns an error if the model cannot be inserted, or if a connection to the database cannot be established.
    async fn create(self) -> Result<Self, query::Error>;

    /// Update the model in the database.
    ///
    /// # Errors
    ///
    /// Returns an error if the model cannot be updated, or if a connection to the database cannot be established.
    async fn save(&mut self) -> Result<(), query::Error> {
        query::save(self).await
    }

    /// Delete the model from the database.
    ///
    /// # Errors
    ///
    /// Returns an error if the model cannot be deleted, or if a connection to the database cannot be established.
    async fn delete(mut self) -> Result<(), query::Error> {
        query::delete(&self).await
    }

    /// Reload a fresh model instance from the database.
    ///
    /// # Errors
    /// Returns an error if the model cannot be retrieved, or if a connection to the database cannot be established.
    async fn fresh(&self) -> Result<Self, query::Error> {
        query::find(self.primary_key()).await
    }

    #[must_use]
    fn query() -> Builder {
        Builder::new(Self::TABLE_NAME.to_string())
    }
}
