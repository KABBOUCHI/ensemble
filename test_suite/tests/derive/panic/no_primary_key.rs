use ensemble::Model;

#[derive(Model)]
struct MyModel {
    name: String,
}

fn main() {}
