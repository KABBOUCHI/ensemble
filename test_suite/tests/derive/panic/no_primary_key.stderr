error: No primary key found. Either mark a field with `#[model(primary)]` or name it `id`.
 --> tests/derive/panic/no_primary_key.rs:4:16
  |
4 |   struct MyModel {
  |  ________________^
5 | |     name: String,
6 | | }
  | |_^
